### PR TITLE
Add failing tests for #1173: off-by-one discrepancy in the location of reported parser errors

### PR DESCRIPTION
**Pull request by hal7df**
This adds (currently failing) test cases for the issue identified in #1173. I've tried my best at adding a plethora of different variations of invalid JSON, but it's possible I might still be missing some edge cases. I also have written this to test every variation of invalid JSON against the four parser backends identified by @cowtowncoder in the issue comments.

### CODE REVIEW

**Comment by hal7df** [comment]
Curiously, `DataInput`-based parsers pass all of these tests, because they don't provide byte-/character-level position information. It's possible that they might still be susceptible to the off-by-one error if the error occurs across a line boundary, but I haven't been able to get that behavior to occur yet.

File: src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+            21
+        ),
+        new InvalidJson(
+            "Invalid JSON with raw unicode character",
+            // javac will parse the unicode control sequence, it will be passed to the parser as a raw unicode character
+            "{\"validJson\":\"\u274c\",\"right\", \"here\"}",
+            26,
+            24,
+            1,
+            25
```
Comment (hal7df) [inline src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java:230]: Parser backends aren't consistent about the column in this case, because the exact definition of "column" seems to depend on whether the parser is operating on a byte array or a character array. I'm open to suggestions about what to do with this case.

Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java:230]: Column should be exactly same for byte/char -backed input, for ASCII characters (for multi-byte characters there is indeed difference). But I think few tests use characters outside ASCII range..

File: src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+        ),
+        ASYNC(
+            (String input) -> {
+                JsonParser parser = JSON_F.createNonBlockingByteArrayParser();
+                ByteArrayFeeder feeder = (ByteArrayFeeder) parser.getNonBlockingInputFeeder();
+                assertTrue(feeder.needMoreInput());
+
+                byte[] inputBytes = input.getBytes(StandardCharsets.UTF_8);
+                feeder.feedInput(inputBytes, 0, inputBytes.length);
+                feeder.endOfInput();
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java:59]: I don't think this works the way you'd expect... I think `endOfInput()` should not be called until all input is consumed.

**Review by cowtowncoder — APPROVED** [review]
Ok parameterization etc make things bit hard to follow but I assume things are fine.
I do have suspicion async-parser case won't work as expected but I'll merge and see how things fare.

File: src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java (hunk 2 above)
Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/failing/read/LocationOfError1173Test.java:59]: Or maybe I misremember this part, as test appears to work fine :)
